lib/rewrite/rewrite-expr-grammar.ym

```diff
@@ -105,7 +105,12 @@ rewrite_expr
             last_rewrite = log_rewrite_unset_new(configuration);
           }
           '(' rewrite_set_opts ')'              { $$ = last_rewrite; }
-	| KW_SET_TAG '(' string ')'             { $$ = log_rewrite_set_tag_new($3, TRUE, configuration); free($3); }
+	| KW_SET_TAG '(' string 
+	  { 
+	    last_rewrite = log_rewrite_set_tag_new($3, TRUE, configuration);
+	    free($3); 
```

Comment:
syslog-ng crashes with the following config:

```
@version: 3.26

log {
  rewrite { set-tag("x" condition()); };
};
```

```
$ ./syslog-ng -Fe -f ../etc/set-tag-invalid.conf 
Error parsing filter expression, syntax error, unexpected ')' in ../etc/set-tag-invalid.conf:4:35-4:36:
1       @version: 3.26
2       
3       log {
4----->   rewrite { set-tag("x" condition()); };
4----->                                   ^
5       };

Segmentation fault (core dumped)
```

The reason is that when there is a lexer error, the lexer itself tries to take care of cleanup, and frees `$3` itself.  As the execution is passed over your block, `$3` is already freed here, so the lexer will do a double free during the cleanup.

To avoid that, you need to do the deallocation on the last token, that is guaranteed not to fail: `')'` in this case.

Reply:
@furiel : I was able to reproduce the crash which you mentioned. I am deallocating last token after `')'` now.